language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/PrimitiveValueDecoders.java

```diff
@@ -116,17 +117,67 @@ private static IllegalArgumentException mismatched(Class<?> clazz) {
    * @hidden
    */
   public static List<com.daml.ledger.javaapi.data.DamlRecord.Field> recordCheck(
-      int expectedFields, Value maybeRecord) {
+      int expectedFields, int optionalFields, Value maybeRecord) {
     var record =
         maybeRecord
             .asRecord()
             .orElseThrow(
                 () -> new IllegalArgumentException("Contracts must be constructed from Records"));
     var fields = record.getFields();
     var numberOfFields = fields.size();
-    if (numberOfFields != expectedFields)
-      throw new IllegalArgumentException(
-          "Expected " + expectedFields + " arguments, got " + numberOfFields);
+    if (numberOfFields == expectedFields) {
+      return fields;
+    }
+    if (numberOfFields > expectedFields) {
+      // Downgrade, check that the additional fields are empty optionals.
+      for (var i = expectedFields; i < numberOfFields; i++) {
+        final var field = fields.get(i);
+        final var optValue = field.getValue().asOptional();
+        if (optValue.isEmpty()) {
+          throw new IllegalArgumentException(
+              "Expected "
+                  + expectedFields
+                  + " arguments, got "
+                  + numberOfFields
+                  + " and field "
+                  + i
+                  + " is not optional: "
+                  + field);
+        }
+        final var value = optValue.get();
+        if (!value.isEmpty()) {
+          throw new IllegalArgumentException(
```

Comment:
These are the same rules for downgrades in the ledger API though. Do we really want to introduce different conversion rules for the java codegen?

Reply:
If I understand your suggestion correctly, it would be an extension of the changes I'm making here in the sense that you'll allow strictly more conversions. If that's the case, how about we merge this in this form to unblock validation of the upgrading design and then add further extensions later if needed?

Reply:
> I'm not sure exactly what this is referring to.

The downgrade rules  in Daml throw an error if you try to use a contract with an optional field that is non-empty in a context that expects an older version (e.g. by exercising a choice on it)

> The situation I'm considering is the that the relevant participants have all vetted a new version of a contract and so compatible new data, e.g. an extra field, will start being produced and populated. We don't want that vetting workflow to necessarily have to be synchronised with upgrading, recompiling and redeploying any java applications which consume this data. As long as the changes are compatible additions, that upgrade could happen on it's own timeline, according to business needs.

That's a fair point. I'm not opposed to adding support for unknown fields separately. That said, when we thought through the scenarios in CN anything that parses history becomes quite problematic if you don't upgrade your java code in sync since you might just miss some choices that really matter.